### PR TITLE
Openapi explain

### LINKED ISSUE #49465: "kubectl explain" should be able to explain "apiservices" and "customresourcedefinition"
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
```
# kubectl explain apiservices
error: group apiregistration.k8s.io has not been registered
```

```
# kubectl explain customresourcedefinition
error: group apiextensions.k8s.io has not been registered
```

**What you expected to happen**:
`kubectl explain` should be able to explain the above two resources. 

Print something as below, etc.
```
DESCRIPTION:
APIService represents a server for a particular GroupVersion. Name must be "version.group".

FIELDS:
......
```

**How to reproduce it (as minimally and precisely as possible)**:
Run the commands above.

**Anything else we need to know?**:
/sig cli
/sig api-machinery

**Environment**:
- Kubernetes version (use `kubectl version`): a9bf44101b3ec13a3e94df926c95ba5384d18e28
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by apelisse**
**What this PR does / why we need it**:
This rewrites `kubectl explain` but using openapi rather than swagger 1.2. Also removes the former code.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #49465, fixes partially #44589, fixes partially #38637

**Special notes for your reviewer**:
FYI @mbohlool 

**Release note**:
```release-note
`kubectl explain` now uses openapi rather than swagger 1.2.
```

### CODE REVIEW

**Comment by pwittrock** [comment]
Reviewed 18 of 23 files at r1.
Review status: 18 of 23 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/kubectl/explain/recursive_fields.go, line 25 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvU_UIs3JwPShDTMdVB:-KvU_UIs3JwPShDTMdVC:ba6sozp) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/recursive_fields.go#L25)):*
> ```Go
> const indentPerLevel = 3
> 
> type recursiveFields struct {
> ```

Document what this is supposed to do

---

*[pkg/kubectl/explain/recursive_fields.go, line 25 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvU_hXbHf2x_1OmvJBa:-KvU_hXcKtZVwtPTSAKL:b-6hqdiy) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/recursive_fields.go#L25)):*
> ```Go
> const indentPerLevel = 3
> 
> type recursiveFields struct {
> ```

How about `recursiveFieldPrinter`

---

*[pkg/kubectl/explain/recursive_fields.go, line 63 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvU_OPKSm-KDkxMH8GJ:-KvU_OPKSm-KDkxMH8GK:bim2rur) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/recursive_fields.go#L63)):*
> ```Go
> }
> 
> func (f *recursiveFields) Print(schema openapi.Schema) error {
> ```

Would you document these functions with what they do.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
Review status: 18 of 23 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubectl/explain/recursive_fields.go, line 23 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvU_pBxEw7vrdltT77P:-KvU_pBxEw7vrdltT77Q:b-fye7hr) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/recursive_fields.go#L23)):*
> ```Go
> )
> 
> const indentPerLevel = 3
> ```

should this be a flag?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
Review status: 18 of 23 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/kubectl/explain/recursive_fields.go, line 37 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvU_sCGnOAMxNrWwBWf:-KvU_sCGnOAMxNrWwBWg:b-368hb8) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/recursive_fields.go#L37)):*
> ```Go
> }
> 
> func (f *recursiveFields) VisitKind(k *openapi.Kind) {
> ```

Document what this do.  I think the linter will complain anyway since they are public.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
Review status: 18 of 23 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/kubectl/explain/recursive_fields_test.go, line 36 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvUa9dXIjExDMvB86Kp:-KvUa9dXIjExDMvB86Kq:biaxgja) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/recursive_fields_test.go#L36)):*
> f
fyi - this reads weird with the indentation being so important.  maybe start on the next line and wrap with a fn to strip the preceeding ws or something?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
Review status: 18 of 23 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/kubectl/explain/recursive_fields_test.go, line 26 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvUaOtkDBG8eP42SFPT:-KvUaOtkDBG8eP42SFPU:btud3mz) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/recursive_fields_test.go#L26)):*
> ```Go
> )
> 
> func TestRecursiveFields(t *testing.T) {
> ```

No gomega?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
Review status: 18 of 23 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/kubectl/explain/typename.go, line 56 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvUaqmSkh66SPzEM9Ez:-KvUaqmSkh66SPzEM9F-:b-v1upud) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/typename.go#L56)):*
> ```Go
> 
> // GetTypeName returns the type of a schema.
> func GetTypeName(schema openapi.Schema) string {
> ```

It is hard to follow the overall code flow with each of these embedded in the Visitors.   Not sure if there is a great way around it though.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
Review status: 18 of 23 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/kubectl/explain/schema_walker.go, line 25 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvUbTThGP1JbT2Goz48:-KvUbTThGP1JbT2Goz49:b-ny9u2t) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/schema_walker.go#L25)):*
> ```Go
> )
> 
> type walker struct {
> ```

call this `lookupField`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
Review status: 18 of 23 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/kubectl/explain/schema_walker.go, line 25 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvUbTThGP1JbT2Goz48:-KvUbkizKdCcFNCmdmY1:b-vj9y8q) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/schema_walker.go#L25)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary><blockquote>

call this `lookupField`
</blockquote></details>

or maybe `lookupFieldForSchema`

---

*[pkg/kubectl/explain/schema_walker.go, line 93 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-KvUbZzT7sHIFVnKEuRx:-KvUbZzT7sHIFVnKEuRy:bqsqth8) ([raw file](https://github.com/kubernetes/kubernetes/blob/d667cb9f3e6282377e51668419ca576098d50045/pkg/kubectl/explain/schema_walker.go#L93)):*
> ```Go
> 
> // FindFieldSchema looks for the schema of a given path in a base schema.
> func FindFieldSchema(schema openapi.Schema, path []string) (openapi.Schema, error) {
> ```

How about `LookupSchemaForField`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228)*

**Comment by pwittrock** [comment]
In general, it might be easier to read if the vistor implementations shared the name of the function used to assess them.  e.g. `PrintRecursive` would instantiate a `printRecursive` struct

---

Review status: 18 of 23 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/53228#-:-KvUbqAGkkVIuS61EWtL:b-fx4uat)*

**Comment by apelisse** [comment]
/retest

@pwittrock I've documented every function and renamed all the things.
